### ISSUE TITLE
As an iOS user I want to reply quickly to an incoming message without unlocking my iPhone

### ISSUE THREAD

**Issue by larsboehnke** [WhisperSystems/Signal-iOS#1406]
This could be implemented by the iOS quick reply functionality like Whats App or Facebook messenger already uses.

**Comment by pjctech** [comment]
+1

**Comment by dreyhomedev** [comment]
+1

**Comment by michaelkirk** [comment]
@dreyhomedev - Please stop bumping issues. It causes subscribers to get spammed. If you'd like to voice your support for an issue, use the github "reaction" feature instead.

**Comment by matth3w** [comment]
Is this not a duplicate of #540 ?

**Comment by lazerwalker** [comment]
Looking at the codebase, it appears to be implemented (https://github.com/WhisperSystems/Signal-iOS/blob/master/Signal/src/network/PushManager.m#L362-L365). 

That being said, this isn't working for me either. My suspicion is that changes in iOS 10 may have broken it, rather than it having never been working?

**Comment by jschneier** [comment]
I can verify this works on iOS 9 (I never upgrade).

**Comment by larsboehnke** [comment]
@matth3w no, #540 is only for conversations while you are in the app. This issue is related to replies from the lock screen.

**Comment by lazerwalker** [comment]
If this is in fact an iOS 10-specific bug, the correct solution is probably to conditionally use the new [UserNotifications](https://developer.apple.com/reference/usernotifications) framework on iOS 10, as UILocalNotification et al are officially deprecated.

BUILDING.md points out that push notifications don't work for third-party developers due to how PN cert signing works on Apple's end. Is there an existing flow for testing push notification-related features as a contributor? Since all user-facing PNs are (I believe?) actually local notifications scheduled in response to silent remote notifications, it should be possible to have test code that schedules local notifications for manual verification.

**Comment by matth3w** [comment]
I used Signal on iOS 9 as well as on 10, and I haven't ever noticed a way to do quick replies on the lock screen, or anywhere else for that matter. Taping or swiping down on a notification has always just taken me to the app. 
@larsboehnke It seems to me, that if it was implemented in one location, such as quick replies within the app, it would work everywhere on platform. That is why I thought it would be considered a duplicate of #540 . I can see why this can be considered a different issue though. I hope both can be implemented.

**Comment by jschneier** [comment]
@matth3w on iOS 9 I swiped left on the notification and the `Reply` button was available to me. I have since upgraded to iOS 10 (what a mistake...) but that is what I was confirming.

**Comment by jschneier** [comment]
I think Apple just capriciously breaks things for fun. Push notifications are totally broken for me on iOS 10, my screen doesn't even wake. I think we need this?

http://ashishkakkad.com/2016/09/push-notifications-in-ios-10-objective-c/

I think that will also fix #1467.

**Comment by lazerwalker** [comment]
@jschnier Yes, I agree completely.

As I suggested when I mentioned that earlier in this thread, someone from the community hopping on this and filing a PR is probably blocked on having a sensible way to test push notifications in the debug build of the app. Maybe @michaelkirk has thoughts?

**Comment by michaelkirk** [comment]
I think it's currently possible, but certainly not simple, to test these notifications. TBH I haven't dove into this part of the codebase myself.

I will leave some notes for the brave:

Keep in mind the actionable notifications you receive aren't actually "push notifications", but rather they are locally generated notifications in response to received voip push events.

I think (?) this is where they're generated: https://github.com/WhisperSystems/Signal-iOS/blob/master/Signal/src/environment/NotificationsManager.m#L102

The other class which may be relevant is https://github.com/WhisperSystems/Signal-iOS/blob/master/Signal/src/network/PushManager.m which is our PKPushRegistryDelegate

So, it should be possible to fake the generation of those notifications using those classes, or something like them. If you come up with a nice way of doing this, maybe they should be included in our test fakes.

Sorry to not be of much help at the moment, but wanted to leave a few breadcrumbs if anyone is interested in pursuing.

**Comment by michaelkirk** [comment]
Oh, and this is where the notifications are *actually* triggered:https://github.com/WhisperSystems/SignalServiceKit/blob/master/src/Messages/TSMessagesManager.m#L594

**Comment by michaelkirk** [comment]
Fixed by https://github.com/WhisperSystems/Signal-iOS/commit/680b2c20d17c8c6a9150461c508254590def0770
